nexus/src/app/sagas/region_replacement_start.rs

```diff
@@ -0,0 +1,1261 @@
+
+    // One the new region was swapped in, create a fake volume record for the
+    // old region record. This will be deleted after region replacement has
+    // finished.
+
+    let volume_construction_request = VolumeConstructionRequest::Volume {
+        id: new_volume_id,
+        block_size: 0,
+        sub_volumes: vec![VolumeConstructionRequest::Region {
+            block_size: 0,
+            blocks_per_extent: 0,
+            extent_count: 0,
+            gen: 0,
+            opts: CrucibleOpts {
+                id: new_volume_id,
+                target: vec![old_region_address.to_string()],
+                lossy: false,
+                flush_timeout: None,
+                key: None,
+                cert_pem: None,
+                key_pem: None,
+                root_cert_pem: None,
+                control: None,
+                read_only: false,
+            },
+        }],
+        read_only_parent: None,
+    };
+
+    let volume_data = serde_json::to_string(&volume_construction_request)
+        .map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&e.to_string()))
+        })?;
+
+    let volume = db::model::Volume::new(new_volume_id, volume_data);
+
+    osagactx
+        .datastore()
+        .volume_create(volume)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+async fn srrs_create_fake_volume_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+
+    // Delete the fake volume.
+
+    let new_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+    osagactx.datastore().volume_hard_delete(new_volume_id).await?;
+
+    Ok(())
+}
+
+async fn srrs_update_request_record(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    let osagactx = sagactx.user_data();
+    let datastore = osagactx.datastore();
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let saga_id = sagactx.lookup::<Uuid>("saga_id")?;
+    let new_dataset_and_region = sagactx
+        .lookup::<(db::model::Dataset, db::model::Region)>(
+            "new_dataset_and_region",
+        )?;
+    let new_region_id = new_dataset_and_region.1.id();
+
+    let old_region_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+
+    // Now that the region has been ensured and the construction request has
+    // been updated, update the replacement request record to 'Running' and
+    // clear the operating saga id. There is no undo step for this, it should
+    // succeed idempotently.
+    datastore
+        .set_region_replacement_running(
+            &opctx,
+            params.request.id,
+            saga_id,
+            new_region_id,
+            old_region_volume_id,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
```

Comment:
From the saga node's perspective, it could either spin until a result is known (retrying within the DataStore function), or fail immediately, unwinding the saga. Though, if Nexus can't talk to CRDB, then unwinding will also fail, and there'll need to be follow up work to address this.